### ISSUE TITLE
How to Cmd+Q to quit app on MacOS?

### ISSUE THREAD

**Issue by AllenDang** [sciter-sdk/go-sciter#142]
How to implement cmd+q to exit the app?

I think this should be supported automatically if I can add a menu, but after a few research (google, sciter forum, etc...) I still cannot find a clue about it.

Any help?

**Comment by pravic** [comment]
Yes, there should be a main menu. It must be described in Apple's documentation: https://developer.apple.com/macos/human-interface-guidelines/menus/menu-bar-menus/#app-menu

**Comment by AllenDang** [comment]
Thanks for the link. So I think you mean this is not implement inside sciter, right?
Ok, I will take a look at cgo.

**Comment by pravic** [comment]
Also note that I am, personally, just a maintainer of bindings, so take my opinion with a grain of salt. Better discuss or ask such things on the official forum, where Andrew @c-smile (the author) answers officially.

**Comment by c-smile** [comment]
This event handler will catch CMD+Q (on Mac) or CTRL+Q (on Win/Lin) and close the window:
```
event keydown (evt) {
   if( evt.shortcutKey && evt.keyCode == Event.VK_Q )
   {
      view.close();
      return true; // consume the event.
   }  
}
```
If that's main window then it will close the application.

**Comment by mchampaneri** [comment]
@pravic 

In go, we have options to integrate c code using cgo. Which already used by go-sciter already.

> To add native mac menu, we should add a os restricted function like addMacMenubar.


Here is small example of menu add to menubar using cgo and cocoa.

[Add menu to mac os app](https://github.com/mchampaneri/go-sciter-example/tree/master/16%20Add%20Menu%20in%20menubar%20in%20MacOS%20) 

**Would like to get your opinion and hopping for positive response.**

**Comment by pravic** [comment]
@mchampaneri Constructing the application menu in runtime - interesting! I thought that it is possible only via [static menu resources](https://github.com/c-smile/sciter-sdk/blob/master/demos/usciter/usciter/usciter/Base.lproj/MainMenu.xib).

Perhaps, I'll take this idea for rust-sciter as well.

**Comment by pravic** [comment]
@mchampaneri by the way, your examples are great! Have you considered placing them under github.com/sciter-sdk organization? You will be the author as usually, but the repository will be more visible for go-sciter users. What do you think?

**Comment by mchampaneri** [comment]
@pravic 
it will be great pleasure for me.

**Comment by c-smile** [comment]
Just an idea: if system menu is required then I think it makes sense to make it as native behaviors. So anyone can define:
```
<head>
    <menu.system>
         <li accesskey="... caption="File">
             <menu>
                  <li accesskey="..." caption="Open" />
                  <li accesskey="... caption="Quit"/>
            </menu>
         </li> 
    </menu>   
</head>
<body>
   ....
</body>
```
And css:
```
head > menu.system { behavior:system-menu; }
```
The system-menu behavior implementation is different for different OSes.

Another option is to provide some native method like `view.defineMenu()`

**Comment by mchampaneri** [comment]
@c-smile 
I tried the way you told. Yes its a feasible one.

```
<script type='text/tiscript'>
                function self.ready(){
                    view.LoadMenu()
                    view.StartApp()
                }
</script>
```

```
win.DefineFunction("LoadMenu", func(...*sciter.Value) *sciter.Value {
		fmt.Println("ok")
		C.LoadMenu()
		return sciter.NewValue("")
	})

win.DefineFunction("StartApp", func(...*sciter.Value) *sciter.Value {
		win.Show()
		return sciter.NewValue("")
	})

```

> There is just one problem I am facing. When app launches first time mac menu does not response to click, but when I reselect the app it works normally. 
 It's not related to view.LoadMenu(). It was also occurred when I called the function from native go code as well.

I

**Comment by mchampaneri** [comment]
@c-smile 

Defining function this way might force to Implement view.defineMenu() for sdk wrappers in every language. 

In sciter-forum, I have seen a post regards python. Some one has written that python can not make c-language call. If python can not make a c-lang call, We will not achieve what we want.

Sciter-sdk itself is in c and so I guess It will be better option if sciter-sdk itself do this, So other wrapper just have to pass on data for runtime menu.

What do you think?

**Comment by pravic** [comment]
@mchampaneri 

> Some one has written that python can not make c-language call. If python can not make a c-lang call, We will not achieve what we want.

Why? Python has [ctypes](https://docs.python.org/3/library/ctypes.html) for cross-language calls: 
https://github.com/sciter-sdk/pysciter/blob/master/sciter/platform.py#L95

**Comment by mchampaneri** [comment]
@pravic 

Thanks for sharing code.  I am not much aware about python.
It seems like sciter already using cocoa. We just have to add menu in case of mac.


@c-smile 

**_This is just an opinion._**

We should go with native menu for **Mac OS X and Ubuntu like Oses** which have **main menu outside application window which needs native support**. 
For windows or other **Oses** where **main menu is within app window** itself **does not require native support.** 


I am agree with you @c-smile . **_We should call Menu binding from view, by that way frontend code of application will be same regardless of Os in context of main menu._**

**Comment by mchampaneri** [comment]
@c-smile 

There is one more requirement to make menu working.
Build has to be inside Contents/MacOS folder. Otherwise it will have this issue.

>There is just one problem I am facing. When app launches first time mac menu does not response to click, but when I reselect the app it works normally.
It's not related to view.LoadMenu(). It was also occurred when I called the function from native go code as well.
